test/gst-msdk/encode/12bit/hevc.py

```diff
@@ -0,0 +1,100 @@
+###
+### Copyright (C) 2023 Intel Corporation
+###
+### SPDX-License-Identifier: BSD-3-Clause
+###
+
+from .....lib import *
+from .....lib.gstreamer.msdk.util import *
+from .....lib.gstreamer.msdk.encoder import EncoderTest
+
+spec = load_test_spec("hevc", "encode", "12bit")
+
+@slash.requires(*have_gst_element("msdkh265dec"))
+@slash.requires(*have_gst_element("msdkh265enc"))
+class HEVC12EncoderBaseTest(EncoderTest):
+  def before(self):
+    super().before()
+    vars(self).update(
+      codec         = "hevc-12",
+      gstdecoder    = "msdkh265dec",
+      gstencoder    = "msdkh265enc",
+      gstmediatype  = "video/x-h265",
+      gstparser     = "h265parse",
+    )
+
+  def get_file_ext(self):
+    return "h265"
+
+
+@slash.requires(*platform.have_caps("encode", "hevc_12"))
+class HEVC12EncoderTest(HEVC12EncoderBaseTest):
+  def before(self):
+    super().before()
+    vars(self).update(
+      caps      = platform.get_caps("encode", "hevc_12"),
+      lowpower  = False,
+    )
+
+class cqp(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, qp, quality, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      case      = case,
+      gop       = gop,
+      profile   = profile,
+      qp        = qp,
+      quality   = quality,
+      rcmode    = "cqp",
+      slices    = slices,
+    )
+
+  @slash.parametrize(*gen_hevc_cqp_parameters(spec, ['main12']))
+  def test(self, case, gop, slices, bframes, qp, quality, profile):
+    self.init(spec, case, gop, slices, bframes, qp, quality, profile)
+    self.encode()
+
+class cbr(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, bitrate, fps, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      bitrate   = bitrate,
+      case      = case,
+      fps       = fps,
+      gop       = gop,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      profile   = profile,
+      rcmode    = "cbr",
+      slices    = slices,
+    )
+
+  @slash.parametrize(*gen_hevc_cbr_parameters(spec, ['main12']))
+  def test(self, case, gop, slices, bframes, bitrate, fps, profile):
+    self.init(spec, case, gop, slices, bframes, bitrate, fps, profile)
+    self.encode()
+
+class vbr(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, bitrate, fps, quality, refs, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      bitrate   = bitrate,
+      case      = case,
+      fps       = fps,
+      gop       = gop,
+      maxrate   = bitrate * 2,
+      minrate   = bitrate,
```

Comment:
Do we need to set maxrate and minrate in the same way for gst-va, gst-vaapi and gst-msdk ?

Reply:
see reply in https://github.com/intel/vaapi-fits/pull/491#discussion_r1160829998

Reply:
However, I think gst-msdk will end up with 50% target-percentage.... which is in line with our ffmpeg default tests settings.  The 70% for gst-vaapi/gst-va was a left-over side-effect of it being hard-coded previously.

Reply:
IIUC, the gst-va refs were copied from gst-vaapi.  In the future, we should make them all use 50% target but will require new PSNR refs.

Reply:
Yes, good point.

That was the original idea we wanted.  Unfortunately, when we wrote the other tests, gst-vaapi used to hard-code target-percentage to 70%.  So it was not possible to make minrate 50% of maxrate in gst-vaapi.  But now gst-vaapi has the property to allow this.  We kept it at 70% for existing tests to avoid rebasing PSNR refs.  And when we copied the old test to make new ones, we should have changed the new ones to be consistent, but it was forgotten :(.

Pretty sure we use 50% in ffmpeg tests, too.

Since these are new tests, I will make them the same.  Thanks.